target/linux/generic/files/drivers/net/phy/rtl8367b.c

```diff
@@ -716,33 +734,46 @@ static int rtl8367b_write_phy_reg(struct rtl8366_smi *smi,
 static int rtl8367b_init_regs(struct rtl8366_smi *smi)
 {
 	const struct rtl8367b_initval *initvals;
+	u32 chip_num;
 	u32 chip_ver;
 	u32 rlvid;
 	int count;
 	int err;
 
 	REG_WR(smi, RTL8367B_RTL_MAGIC_ID_REG, RTL8367B_RTL_MAGIC_ID_VAL);
+	REG_RD(smi, RTL8367B_CHIP_NUMBER_REG, &chip_num);
 	REG_RD(smi, RTL8367B_CHIP_VER_REG, &chip_ver);
 
 	rlvid = (chip_ver >> RTL8367B_CHIP_VER_RLVID_SHIFT) &
 		RTL8367B_CHIP_VER_RLVID_MASK;
 
-	switch (rlvid) {
-	case 0:
-		initvals = rtl8367r_vb_initvals_0;
-		count = ARRAY_SIZE(rtl8367r_vb_initvals_0);
-		break;
+	if (smi->flags & RTL8367B_FLAG_MATCH_CHIP_NUM) {
```

Comment:
If we are introducing the chip_id match because of rtl8367c family, why not use that instead of using a flag? All devics using the compatible realtek,rtl8367c should use the chip_id match.

Edit: oh... flag came from the match data pointer. It is not pretty to use a pointer as a casted u32 bitmap (which will break with larger pointer size). It is way better to use a const struct with all required info in a readable form. That's the idea of the matching data.

Anyway, we are detecting the chip twice (detect and initregs). It should have save that info and just reuse it here.

Reply:
I was hoping that it would be possible to base rtl8167d support on this code. In the case of the Tp-link router with Archer C5 v4 which comes in two versions rtl8167c and rtl8167d. If set compatible to realtek, rtl8367c will not work rtl8367d. For this reason, I wanted further detection to occur within the realtek,rtl8367c